pandas/io/parsers.py

```diff
@@ -1685,7 +1685,22 @@ def _cast_types(self, values, cast_type, column):
 
         else:
             try:
-                values = astype_nansafe(values, cast_type, copy=True)
+                # gh-20377
+                #
+                # The C parser does not convert np.NaN to 'nan' if
+                # the casted dtype is a string-like.
+                cast_type = pandas_dtype(cast_type)
+
```

Comment:
woa, this is crazy logic here, is there a reason you don't just push this to astype_nansafe itself?

Reply:
It's actually not that crazy as it looks.  It mirrors partially what's in `astype_nansafe` but localizes the patch to just the string types for this function.  This is what I was referring to when I said I wanted to localize the change to just `parsers.py`.

Alternatively, I could put `skipna=False` as a new parameter to `astype_nansafe`, but then that patch has much greater reaching effects than I was intending (which was my concern about #20429, especially since you had mentioned that we wanted to preserve the behavior of casting `np.NaN` to `'nan'` in some cases).

Reply:
> you had mentioned that we wanted to preserve the behavior of casting np.NaN to 'nan' in some cases)

we *never* want to do this, maybe my comment was confusing.

I would limit the changes to just adding a parameter to ``astype_nansafe``

Reply:
Gotcha.  Done.